_posts/2020-12-27-spring-boot-exception-hanlding.md

```diff
@@ -0,0 +1,596 @@
+We might want to keep the value of `include-stacktrace` flag to `never` only, at least in production, as it might reveal internal
+workings of the application.
+
+Moving on! The status and error message - `500` - indicates that something is wrong with our code but in actuality, it's a client error, and
+our current status code doesn't correctly reflect that. Unfortunately, this is as far as we can go with the `server.error` configurations.
+
+## `@ResponseStatus`
+
+`@ResponseStatus` as the name suggests allows us to modify the HTTP status of our response. It can be applied at following
+places:
+* On the exception class itself
+* Along with `@ExceptionHandler` annotation on methods
+* Along with `@ControllerAdvice` annotation on classes
+
+In this section, we will be looking at the first point only.
+
+Let's come back to the problem at hand which is our error responses aren't giving us status codes that better represent the
+corresponding error message.
+
+To address this we can we annotate our Exception class with `@ResponseStatus` and pass in the desired HTTP response status
+in its value property:
+
+```java
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class NoSuchElementFound extends RuntimeException {
+
+}
+```
+
+This change will result in a much better response:
+
+```json
+{
+    "timestamp": "2020-11-29T09:42:12.287+00:00",
+    "status": 404,
+    "error": "Not Found",
+    "message": "Item with id 1 not found",
+    "path": "/product/1"
+} 
+```
+
+Another way to achieve the same is by extending `ResponseStatusException` class:
+
+```java
+public class NoSuchElementFoundException extends ResponseStatusException {
+
+    public NoSuchElementFoundException(String message){
+        super(HttpStatus.NOT_FOUND, message);
+    }
+}
+```
+
+This approach comes in handy when we want to manipulate the response headers too.
+
+Now, as we can see `@ResponseStatus` along with `server.error` properties allows us to manipulate almost all the fields
+in our Spring defined error response payload but, what if want to manipulate response payload structure too? Let's see how
+we can achieve that in the next section. 
+
+## `@ExceptionHandler`
+`@ExceptionHandler` annotation gives us a lot of flexibility in terms of handling exceptions. For starters, to use it, we
+simply need to create a method either in the controller itself or in a `@ControllerAdvice` class and
+annotate it with `@ExceptionHandler`.
+
+```java
+@RestController
+@RequestMapping("/product")
+public class ProductController {
+    private final ProductService productService;
+    //constructor omitted for brevity...
+
+    @GetMapping("/{id}")
+    public Response getProduct(@PathVariable String id) {
+        return productService.getProduct(id);
+    }
+
+    @ExceptionHandler(NoSuchElementFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ResponseEntity<String> handleNoSuchElementFoundException(
+            NoSuchElementFoundException exception
```

Comment:
In some cases you call the exception `NoSuchElementFound` and in others `NoSuchElementFoundException`. Please use only one.